### PR TITLE
Is/shared project concurrency - backend

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?

Adds the concept of a "locked" project. A project is locked once a user opens it.
This is the first attempt concurrent access of a shared project instance. locking of a project is the simplest. As long as a user keeps the lock, another user cannot open the project.

Now when a GUI user successfully logs in and then connects to the websocket. The backend will put that client into as many socket.io "rooms" as the user is in groups. e.g. if the user is part of the groups ```it'is``` and ```sparc```, then he will be in the rooms corresponding to his primary group id, the it'is group id and the sparc group id.

- adds endpoint /projects/{uuid}/state to get a project state (currently let the client know whether a project is locked)
- adds projectStateUpdated socket.io event that transmits the exact same information to connected clients

Bonus:
- small fixes: renamed access_rights to accessRights to follow conventions in the group endpoints (frontend,backend)
- additional tests on groups endpoint with regard to access rights
- adds an initial mypy.ini configuration accessible in services/packages through ```make mypy``` (note: this is still in progress and not automatized in the CI. please test it on your own and see how many issues appear)

## Checklist

- [ ] Did you change any service's API? Then make sure to bundle document and upgrade version (``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [x] Unit tests for the changes exist
- [x] Runs in the swarm
- [ ] Documentation reflects the changes
- [ ] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)

### CODE REVIEW

File: scripts/common.Makefile [hunk 1]
```diff
@@ -100,6 +100,9 @@ autoformat: ## runs black python formatter on this service's code. Use AFTER mak
 		--exclude "/(\.eggs|\.git|\.hg|\.mypy_cache|\.nox|\.tox|\.venv|\.svn|_build|buck-out|build|dist|migration|client-sdk|generated_code)/" \
 		$(CURDIR)
 
+.PHONY: mypy
+mypy: $(REPO_BASE_DIR)/mypy.ini ## runs mypy python static type checker on this services's code. Use AFTER make install-*
+	docker run -ti -v $(REPO_BASE_DIR)/mypy.ini:/config/mypy.ini -v $(CURDIR):/src --workdir=/src kiwicom/mypy mypy --config-file /config/mypy.ini src
```
Comment (GitHK) [inline scripts/common.Makefile:105]: 👍

File: services/web/server/src/simcore_service_webserver/projects/projects_handlers.py [hunk 2]
```diff
@@ -313,28 +328,59 @@ async def close_project(request: web.Request) -> web.Response:
...
+    with managed_resource(user_id, None, request.app) as rt:
+        # TODO: temporary hidden until get_handlers_from_namespace refactor to seek marked functions instead!
+        from .projects_api import get_project_for_user
+
+        await get_project_for_user(
+            request.app,
+            project_uuid=project_uuid,
+            user_id=user_id,
+            include_templates=True,
+        )
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/projects/projects_handlers.py:378]: why is this just called but the result not used? A comment might help to clarify.

File: services/web/server/src/simcore_service_webserver/projects/projects_handlers.py [hunk 3]
```diff
@@ -313,28 +335,70 @@ async def close_project(request: web.Request) -> web.Response:
...
-            other_users = await rt.find_users_of_resource("project_id", project_uuid)
-            if not other_users:
-                # only remove the services if no one else is using them now
-                fire_and_forget_task(
-                    projects_api.remove_project_interactive_services(
+            project_opened_by_others = (
+                len(await rt.find_users_of_resource("project_id", project_uuid)) > 0
+            )
+        # if we are the only user left we can safely remove the services
+        async def close_project_task() -> None:
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/projects/projects_handlers.py:351]: MINOR: internal functions start with ``_``??

File: services/web/server/tests/unit/with_dbs/_helpers.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+from typing import List, Tuple
+
+from aiohttp import web
+
+from simcore_service_webserver.projects.projects_handlers import HTTPLocked
+from simcore_service_webserver.security_roles import UserRole
+
+ExpectedResponse = namedtuple(
+    "ExpectedResponse",
+    ["ok", "created", "no_content", "not_found", "forbidden", "locked"],
```
Comment (pcrespov) [inline services/web/server/tests/unit/with_dbs/_helpers.py:12]: MINOR: I would rename this as ``if_ok``, ``if_created``, ...

File: services/web/server/tests/unit/with_dbs/_helpers.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+)
+
+
+def standard_role_response() -> Tuple[str, List[Tuple[UserRole, ExpectedResponse]]]:
+    return (
```
Comment (pcrespov) [inline services/web/server/tests/unit/with_dbs/_helpers.py:17]: BTW: how did you make it work??

File: services/web/server/tests/unit/with_dbs/conftest.py [hunk 6]
```diff
@@ -239,41 +239,56 @@ async def redis_client(loop, redis_service):
...
 @pytest.fixture()
-async def security_cookie(client) -> str:
-    # get the cookie by calling the root entrypoint
-    resp = await client.get("/v0/")
-    payload = await resp.json()
-    assert resp.status == 200, str(payload)
-    data, error = unwrap_envelope(payload)
-    assert data
-    assert not error
+async def security_cookie(client) -> Callable:
```
Comment (pcrespov) [inline services/web/server/tests/unit/with_dbs/conftest.py:251]: CONVENTION: can we agree that fixtures that return a creator callable are named ``factory``?, .e.g. ``security_cookie_factory`` in this case (as ``tmpdir_factory``, etc)

File: services/web/server/tests/unit/with_dbs/conftest.py [hunk 7]
```diff
@@ -239,41 +239,56 @@ async def redis_client(loop, redis_service):
...
+    async def creator(client_override: Optional = None) -> str:
```
Comment (pcrespov) [inline services/web/server/tests/unit/with_dbs/conftest.py:252]: Can you write ``Optional`` without ``[]``? Is this ``Optional[Any]``?? well, i should RTFM :-)

File: services/web/server/tests/unit/with_dbs/conftest.py [hunk 8]
```diff
@@ -239,41 +239,56 @@ async def redis_client(loop, redis_service):
...
+        # get the cookie by calling the root entrypoint
+        resp = await (client_override or client).get("/v0/")
```
Comment (pcrespov) [inline services/web/server/tests/unit/with_dbs/conftest.py:254]: Let's start passing the version as a constant or a fixture... the day we move to ``v1`` it's going to be epic! :-)
I started creating a global fixture called ``def version_tag()`` that returns this number. IN the future we can even create a parametrized fixture that can eventually test multiple active versions

File: services/web/server/tests/unit/with_dbs/conftest.py [hunk 9]
```diff
@@ -239,41 +239,56 @@ async def redis_client(loop, redis_service):
...
+        payload = await resp.json()
```
Comment (pcrespov) [inline services/web/server/tests/unit/with_dbs/conftest.py:255]: why not using ``assert_response`` from ``pytest-simcore`` helpers?

File: services/web/server/src/simcore_service_webserver/projects/projects_handlers.py (hunk 2 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/projects/projects_handlers.py:385]: yep, this is something copy/pasted all over the place... it is basically to check if the project exists. will make it more clear.

File: scripts/common.Makefile (hunk 1 above)
Comment (sanderegg) [reply scripts/common.Makefile:105]: will do. I was probably dreaming of a makefile-only world... ;)

File: services/web/server/src/simcore_service_webserver/projects/projects_handlers.py [hunk 10]
```diff
@@ -291,10 +298,25 @@ async def open_project(request: web.Request) -> web.Response:
                 user_id=user_id,
                 include_templates=True,
             )
+
+            # let's check if that project is already opened by someone else
+            other_users = await rt.find_users_of_resource("project_id", project_uuid)
+            if other_users and any(f"{user_id}" != user for user in other_users):
+                # project is already locked
+                raise HTTPLocked(reason="Project is already opened by another user")
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/projects/projects_handlers.py:306]: will add it

File: services/web/server/tests/unit/with_dbs/_helpers.py (hunk 5 above)
Comment (sanderegg) [reply services/web/server/tests/unit/with_dbs/_helpers.py:17]: some bloody space/parenthis or so I guess...

File: services/web/server/tests/unit/with_dbs/conftest.py (hunk 7 above)
Comment (sanderegg) [reply services/web/server/tests/unit/with_dbs/conftest.py:252]: I guess this is Optional[Any] yes... I think all of these types can be written without type actually.

**Comment by sanderegg** [comment]
> If I understand this correctly. After a user logs in, the backend subscribes him to X amount of channels (which are related to his groups) where he will receive updates on project status changes?
> 
> To know if a project is locked, the currently opened browser tabs are just counted? In theory this approach should work, we'll solve concurrency issues in a later stage, if they occur.

@GitHK , to know if a project is locked the backend asks redis to check if the project is already in use by some user.

File: services/web/server/src/simcore_service_webserver/projects/projects_handlers.py (hunk 3 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/projects/projects_handlers.py:359]: done

File: services/web/server/tests/unit/with_dbs/conftest.py (hunk 6 above)
Comment (sanderegg) [reply services/web/server/tests/unit/with_dbs/conftest.py:252]: done

File: services/web/server/tests/unit/with_dbs/conftest.py (hunk 9 above)
Comment (sanderegg) [reply services/web/server/tests/unit/with_dbs/conftest.py]: done

**Pull request merged**